File: src/main/java/com/mattchowning/file_read_write/client/FileReadWriteClient.java
package com.mattchowning.file_read_write.client;

import com.mattchowning.file_read_write.client.calls.*;
import com.mattchowning.file_read_write.client.handler.HandlerCallback;
import com.mattchowning.file_read_write.server.model.OAuthToken;

import org.jetbrains.annotations.NotNull;

public class FileReadWriteClient {

    private OAuthToken oAuthToken;

    void logout() {
        oAuthToken = null;
    }
    boolean isAuthorized() {
        return oAuthToken != null;
    }

    void getFileContent(HandlerCallback<String> callback) {
        Call<String> call = new GetFileCall(oAuthToken, this, callback);
        call.execute();
    }

    void updateFileContent(@NotNull String newFileContent, HandlerCallback<String> callback) {
        Call<String> call = new PostFileCall(oAuthToken, newFileContent, this, callback);
        call.execute();
    }

    void getOAuthToken(HandlerCallback<OAuthToken> callback, String username, String password) {
        Call<OAuthToken> call = new GetOAuthCall(username,
                                                 password,
                                                 setOAuthTokenWithCallback(callback));
        call.execute();
    }

    public void refreshOAuthToken(HandlerCallback<OAuthToken> callback) {
        Call<OAuthToken> call = new RefreshOAuthCall(oAuthToken.getRefreshToken(),
                                                     setOAuthTokenWithCallback(callback));
        call.execute();
    }

    private HandlerCallback<OAuthToken> setOAuthTokenWithCallback(HandlerCallback<OAuthToken> originatingCallback) {
        return new HandlerCallback<OAuthToken>() {
            @Override
            public void onSuccess(OAuthToken result) {
                FileReadWriteClient.this.oAuthToken = result;
                originatingCallback.onSuccess(result);

            }

            @Override
            public void onError() {
                originatingCallback.onError();
            }
        };
    }
}

File: src/main/scala/com/mattchowning_scala/file_read_write/client/FileReadWriteClient.scala
package com.mattchowning_scala.file_read_write.client

import com.mattchowning_scala.file_read_write.server.model.OAuthToken
import com.mattchowning_scala.file_read_write.client.calls._

class FileReadWriteClient {

  private var oAuthToken: OAuthToken = null

  private val setOAuthToken: (OAuthToken => OAuthToken) = (token: OAuthToken) => {
    this.oAuthToken = token
    token
  }

  def retrieveFileContent(successCallback: String => Unit, failureCallback: () => Unit) {
    val call: Call[String] = new GetFileCall(oAuthToken, this, successCallback, failureCallback)
    call.execute()
  }

  def updateFileContent(newFileContent: String,
                        successCallback: String => Unit,
                        failureCallback: () => Unit) {
    val call: Call[String] = new PostFileCall(oAuthToken, this, successCallback, failureCallback, newFileContent)
    call.execute()
  }

  def retrieveOAuthToken(successCallback: OAuthToken => Unit,
                         failureCallback: () => Unit,
                         username: String,
                         password: String) {
    val call: Call[OAuthToken] = new GetOAuthCall(username,
                                                  password,
                                                  setOAuthToken.andThen(successCallback),
                                                  failureCallback)
    call.execute()
  }

  def refreshOAuthToken(successCallback: OAuthToken => Unit, failureCallback: () => Unit) {
    val call: Call[OAuthToken] = new RefreshOAuthCall(oAuthToken.refreshToken,
                                                      setOAuthToken.andThen(successCallback),
                                                      failureCallback)
    call.execute()
  }
}